### PR TITLE
Add unit tests for io.moquette.spi.impl.security.AES

### PR DESCRIPTION
**Pull request by EricHetti**
I've analysed your codebase and noticed that `io.moquette.spi.impl.security.AES` is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import static org.mockito.Matchers.isNull;
+
+@RunWith(PowerMockRunner.class)
+@PowerMockIgnore({"javax.management.*"})
+public class AESTest {
+
+    @PrepareForTest({Cipher.class, AES.class, System.class})
+    @Test
+    public void testAESEncrypt() throws Exception {
+        PowerMockito.mockStatic(System.class);
```
Comment (paulkennethkent) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Is mocking the `System.class` necessary for this method?

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        Method getInstanceMethod =
+            DTUMemberMatcher.method(Cipher.class, "getInstance", String.class);
+        PowerMockito.doReturn(cipher)
+            .when(Cipher.class, getInstanceMethod)
+            .withArguments(or(isA(String.class), isNull(String.class)));
+
+        Assert.assertNull(AES.AESDecrypt(null, (byte[]) null, false));
+    }
+
+    @PrepareForTest({Cipher.class, AES.class, System.class})
```
Comment (paulkennethkent) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Is the `System.class` required for this test?

**Review by paulkennethkent — CHANGES_REQUESTED** [review]

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 1 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Removed

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 2 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Removed

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        PowerMockito.doReturn(cipher)
+            .when(Cipher.class, getInstanceMethod)
+            .withArguments(or(isA(String.class), isNull(String.class)));
+
+        Assert.assertNull(AES.AESEncrypt("", "0x000x110x220x33"));
+        Assert.assertNull(AES.AESEncrypt(new byte[]{1, 2}, (byte[]) null));
+        Assert.assertNull(AES.AESEncrypt(new byte[]{1, 2}, new byte[]{0x00}));
+    }
+
+    @org.junit.Test
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: `@org.junit.Test` should be just `@Test`

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    public void testGetUnsignedByte() {
+        Assert.assertEquals(12, AES.getUnsignedByte((byte) 12));
+        Assert.assertEquals(210, AES.getUnsignedByte((byte) 1234));
+    }
+
+    @org.junit.Test
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: `@org.junit.Test` should be just `@Test`

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    public void testAESDecrypt1() {
+        byte[] aesKey = new byte[]{0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66,
+            0x77, 0x78, 0x79, 0x7A, 0x7B, 0x7C, 0x7D, 0x7E, 0x7F};
+
+        Assert.assertNull(AES.AESDecrypt(aesKey, "0x000x110x220x33", true));
+        Assert.assertNull(AES.AESDecrypt(new byte[]{1, 2}, new byte[]{0x00}, true));
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        PowerMockito.doReturn(myByteArray)
+            .when(cipher, doFinalMethod)
+            .withArguments(or(isA(byte[].class), isNull(byte[].class)));
+
+        Method getInstanceMethod =
+            DTUMemberMatcher.method(Cipher.class, "getInstance", String.class);
+        PowerMockito.doReturn(cipher)
+            .when(Cipher.class, getInstanceMethod)
+            .withArguments(or(isA(String.class), isNull(String.class)));
+        PowerMockito.when(System.currentTimeMillis()).thenReturn(1_515_585_600_000L);
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        Assert.assertNull(AES.AESDecrypt(new byte[]{1, 2}, new byte[]{0x00}, true));
+    }
+
+    @PrepareForTest({Cipher.class, AES.class, System.class})
+    @Test
+    public void testAESDecrypt2() throws Exception {
+        PowerMockito.mockStatic(System.class);
+        PowerMockito.mockStatic(Cipher.class);
+        Cipher cipher = PowerMockito.mock(Cipher.class);
+        byte[] myByteArray = {(byte) 0, (byte) 0, (byte) 0, (byte) -32, (byte) 0};
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        Assert.assertNull(AES.AESDecrypt(null, (byte[]) null, true));
+    }
+
+    @PrepareForTest({Cipher.class, AES.class, System.class})
+    @Test
+    public void testAESDecrypt3() throws Exception {
+        PowerMockito.mockStatic(System.class);
+        PowerMockito.mockStatic(Cipher.class);
+        Cipher cipher = PowerMockito.mock(Cipher.class);
+        byte[] myByteArray = {(byte) 0, (byte) 0, (byte) 0, (byte) 1, (byte) 0};
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+        Method doFinalMethod =
+            DTUMemberMatcher.method(Cipher.class, "doFinal", byte[].class);
+        PowerMockito.doReturn(myByteArray)
+            .when(cipher, doFinalMethod)
+            .withArguments(or(isA(byte[].class), isNull(byte[].class)));
+
+        Method getInstanceMethod =
+            DTUMemberMatcher.method(Cipher.class, "getInstance", String.class);
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        PowerMockito.doReturn(cipher)
+            .when(Cipher.class, getInstanceMethod)
+            .withArguments(or(isA(String.class), isNull(String.class)));
+        PowerMockito.when(System.currentTimeMillis()).thenReturn(1_515_585_600_000L);
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        byte[] myByteArray = {(byte) 0, (byte) 0, (byte) 0, (byte) 0};
+
+        Method doFinalMethod =
+            DTUMemberMatcher.method(Cipher.class, "doFinal", byte[].class);
+        PowerMockito.doReturn(myByteArray)
+            .when(cipher, doFinalMethod)
+            .withArguments(or(isA(byte[].class), isNull(byte[].class)));
+
+        Method getInstanceMethod =
+            DTUMemberMatcher.method(Cipher.class, "getInstance", String.class);
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        Assert.assertNull(AES.AESDecrypt(null, (byte[]) null, false));
+    }
+
+    @PrepareForTest({Cipher.class, AES.class})
+    @Test
+    public void testAESDecrypt5() throws Exception {
+        PowerMockito.mockStatic(Cipher.class);
+        Cipher cipher = PowerMockito.mock(Cipher.class);
+        Method getInstanceMethod =
+            DTUMemberMatcher.method(Cipher.class, "getInstance", String.class);
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: line > 80 chars

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        PowerMockito.doReturn(cipher)
+            .when(Cipher.class, getInstanceMethod)
+            .withArguments(or(isA(String.class), isNull(String.class)));
+
+        byte[] aesKey = {(byte) 0, (byte) 0, (byte) 0, (byte) 0, (byte) 0, (byte) 0, (byte) 0, (byte) 0,
```
Comment (petersear) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: 2 lines > 80 chars

File: pom.xml [hunk 14]
```diff
@@ -38,6 +38,31 @@
         <module>client</module>
     </modules>
 
+    <dependencies>
```
Comment (petersear) [inline pom.xml]: The correct versions for these dependencies can be found at:
https://github.com/diffblue/platform/blob/develop/docs/user/topics/executing-diffblue-tests.md#maven-projects

**Review by petersear — CHANGES_REQUESTED** [review]

**Review by paulkennethkent — CHANGES_REQUESTED** [review]
I missed this the first time around but there are few instances where the test suggested mocking the class you are testing. I'm not sure if this is needed in the cases below. 

 @PrepareForTest({Cipher.class, **AES.class}**)

**Review by JohnLBergqvist — CHANGES_REQUESTED** [review]
I'm not sure how Cipher works, but does it need to use mocking? If it takes some sort of random seed from the current time, can't you just mock the call to `System.CurrentTimeMillis` instead?

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+import org.powermock.modules.junit4.PowerMockRunner;
+import java.lang.reflect.Method;
+import javax.crypto.Cipher;
+
+import static org.mockito.AdditionalMatchers.or;
+import static org.mockito.Matchers.isA;
+import static org.mockito.Matchers.isNull;
+
+@RunWith(PowerMockRunner.class)
+@PowerMockIgnore({"javax.management.*"})
```
Comment (paulkennethkent) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Potentially can be more specific here if it's required `@PowerMockIgnore({"javax.management. Cipher"})` but at the same time why would we be ignoring something we are trying to mock? I think this needs further investigation.

**Review by paulkennethkent — CHANGES_REQUESTED** [review]

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 13 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 12 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 3 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 4 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 5 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 6 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 10 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 7 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 11 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 8 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 9 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: pom.xml (hunk 14 above)
Comment (Braavos6) [reply pom.xml]: Changed versions to 1.9.0 and 1.6.6

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        PowerMockito.mockStatic(Cipher.class);
+
+        Cipher cipher = PowerMockito.mock(Cipher.class);
+        Method getInstanceMethod =
+            DTUMemberMatcher.method(Cipher.class, "getInstance", String.class);
+        PowerMockito.doReturn(cipher)
+            .when(Cipher.class, getInstanceMethod)
+            .withArguments(or(isA(String.class), isNull(String.class)));
+
+        Assert.assertNull(AES.AESEncrypt("", "0x000x110x220x33"));
```
Comment (freddytuxworth) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: I think all these tests would work if you just remove all of the mocking (you will need to change the expected values though).

**Comment by freddytuxworth** [comment]
@Braavos96 I had a quick look through `Cipher` (and its dependencies `SecureRandom` and `JceSecurity`), and I got the impression that these methods will always be pretty ugly to test. `Cipher` and `SecureRandom` are both non-deterministic in a way that's not easy to "sub out" for something reproducible (they read from `/dev/random`). Also, the `AES` class itself salts the data with `currentTimeMillis()` so that will have to be mocked as well.

One test that you could write for these methods without having to worry about any mocking would be to encrypt and then decrypt a few different pieces of data, and test that the result is the same as the initial data. Not very realistic for what Cover would do though...

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
+package io.moquette.spi.impl.security;
+
+import org.junit.Test;
+import org.junit.Assert;
+
+public class AESTest {
+
+    @Test
+    public void testAESEncrypt() {
+        byte[] aesKey = new byte[]{-30, -41, -27, -46, 40,
```
Comment (freddytuxworth) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Rename this variable `encrypted`

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+            -37, 67, 35, 123, 49, -111, -57, 56, 99, -115, -37};
+
+        Assert.assertArrayEquals(aesKey,
+            AES.AESEncrypt("", "0x000x110x220x33"));
```
Comment (freddytuxworth) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Make `sSrc` non-empty

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+        Assert.assertNull(AES.AESDecrypt(encrypted1,
+            "0x000x110x220x33", true));
+        Assert.assertNull(AES.AESDecrypt(new byte[]{1, 2},
+            "0x000x110x220x33", true));
+        Assert.assertNull(AES.AESDecrypt(null, (byte[]) null, true));
+        Assert.assertNull(AES.AESDecrypt(new byte[]{1, 2},
+            new byte[]{0x00}, true));
+
+        byte[] encrypted2 = AES.AESEncrypt("foobar", "0x000x110x220x33");
+        byte[] aesKey = new byte[]{102, 111, 111, 98, 97, 114};
```
Comment (freddytuxworth) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: rename `aesKey` `foobarBytes`

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 19 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 18 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 17 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Done

**Review by paulkennethkent — APPROVED** [review]

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
+package io.moquette.spi.impl.security;
+
+import org.junit.Test;
+import org.junit.Assert;
+
+public class AESTest {
+
+    @Test
+    public void testEncrypted() {
+        byte[] aesKey = new byte[]{-31, 117, 20, -91, 78,
```
Comment (freddytuxworth) [inline client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Rename this variable `encrypted`

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 20 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: I thought you meant the method name :)
Done

File: client/src/test/java/io/moquette/spi/impl/security/AESTest.java (hunk 15 above)
Comment (Braavos6) [reply client/src/test/java/io/moquette/spi/impl/security/AESTest.java]: Mocking has been removed

**Review by freddytuxworth — APPROVED** [review]

**Review by JohnLBergqvist — APPROVED** [review]

**Review by petersear — APPROVED** [review]

**Pull request closed without merging**